samza-api/src/main/java/org/apache/samza/operators/descriptors/base/system/ExpandingInputDescriptorProvider.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.descriptors.base.system;
+
+import org.apache.samza.operators.descriptors.base.stream.InputDescriptor;
+import org.apache.samza.serializers.Serde;
+
+/**
+ * Interface for advanced {@code SystemDescriptors} that constrain the type of returned {@code InputDescriptors} to
+ * their own {@code StreamExpander} function result types.
+ *
+ * @param <StreamExpanderType> type of the system level {@code StreamExpander} results
+ */
+public interface ExpandingInputDescriptorProvider<StreamExpanderType> {
+
+  /**
+   * Gets a {@link InputDescriptor} for an input stream on this system. The stream has the provided
+   * stream level serde, and the default system level {@code StreamExpander}
+   * <p>
+   * The type of messages in the stream is the type of messages returned by the default system level
+   * {@code StreamExpander}
+   *
+   * @param streamId id of the input stream
+   * @param serde stream level serde for the input stream
```

Comment:
Is this serde applied to all expanded physical streams? This is still a bit fuzzy to me. Here is the sequence of functions that could be applied to the incoming messages: serde used to deserialize the message from physical streams => InputTransformer used to transform the deserialized IncomingMessageEnvelope to some format => StreamExpander that defines a DAG transforming the transformed messages from InputTransformer. So, is this serde the same one that are applied to all different physical streams in the first step? What happens if the expanded DAG needs to read from different physical systems and we need to use different serdes to deserialize the physical streams?

Reply:
It's really up the the system implementers to determine. We will call their expand function with whatever input descriptor the user obtained from their system descriptor and gave to us. They are responsible for creating the sub-dag correctly based on this user-provided input descriptor, including any new input/output descriptors describing the underlying physical streams. For example, they may choose to pass through the user-provided serde, or use their own serde for them. In fact, since this interface is just a guideline, they may even expose a getInputDescriptor(String) version without a user-provided serde and control the serde themselves. Does that answer your question?